testing/trino-testing/src/main/java/io/trino/testing/containers/BaseTestContainer.java

```diff
@@ -0,0 +1,200 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testing.containers;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.net.HostAndPort;
+import io.airlift.log.Logger;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.startupcheck.IsRunningStartupCheckStrategy;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.MountableFile;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.net.HostAndPort.fromParts;
+import static java.util.Objects.requireNonNull;
+
+public abstract class BaseTestContainer
+        implements Closeable, AutoCloseable
+{
+    private final Logger log;
+
+    private final String hostName;
+    private final Set<Integer> ports;
+    private final Map<String, String> filesToMount;
+    private final Map<String, String> envVars;
+    private final Optional<Network> network;
+    private final int startupRetryLimit;
+
+    private GenericContainer<?> container;
+
+    protected BaseTestContainer(
+            String image,
+            String hostName,
+            Set<Integer> ports,
+            Map<String, String> filesToMount,
+            Map<String, String> envVars,
+            Optional<Network> network,
+            int startupRetryLimit)
+    {
+        checkArgument(startupRetryLimit > 0, "startupRetryLimit needs to be greater or equal to 0");
+        this.log = Logger.get(this.getClass());
+        this.container = new GenericContainer<>(requireNonNull(image, "image is null"));
+        this.ports = requireNonNull(ports, "ports is null");
+        this.hostName = requireNonNull(hostName, "hostName is null");
+        this.filesToMount = requireNonNull(filesToMount, "filesToMount is null");
+        this.envVars = requireNonNull(envVars, "envVars is null");
+        this.network = requireNonNull(network, "network is null");
+        this.startupRetryLimit = startupRetryLimit;
+        setupContainer();
+    }
+
+    protected void setupContainer()
+    {
+        for (int port : this.ports) {
+            this.container.addExposedPort(port);
+        }
+        this.filesToMount.forEach(this::copyFileToContainer);
+        this.container.withEnv(this.envVars);
+        this.container.withCreateContainerCmdModifier(c -> c.withHostName(hostName))
+                .withStartupCheckStrategy(new IsRunningStartupCheckStrategy())
+                .waitingFor(Wait.forListeningPort())
+                .withStartupTimeout(Duration.ofMinutes(5));
```

Comment:
It would be useful to have these two configurable without having to override the entire method. Might be out of scope for this PR though.

Reply:
I've approach to it with - Make it configurable once it's required. So I would apply change once it's required in other usecases :)